clio-server/src/main/scala/org/broadinstitute/clio/server/service/ServerService.scala

```diff
@@ -55,26 +54,46 @@ class ServerService private (
     */
   private[service] def recoverMetadata(
     implicit index: ElasticsearchIndex[_]
-  ): Future[Int] = {
+  ): Future[Unit] = {
+    val name = index.indexName
+
     searchDAO.getMostRecentDocument.flatMap { mostRecent =>
-      val msg = s"Recovering all upserts from index ${index.indexName}"
+      val msg = s"Recovering all upserts from index $name"
       val latestUpsert = mostRecent.map(getUpsertId)
 
       logger.info(latestUpsert.fold(msg)(id => s"$msg since ${id.id}"))
 
       persistenceDAO
         .getAllSince(latestUpsert)
-        .runWith(Sink.foldAsync(0) { (count, json) =>
-          logger.debug(s"Recovering document with ID ${getUpsertId(json)}")
-          searchDAO.updateMetadata(json).map(_ => count + 1)
-        })
-        .andThen {
-          case Success(count) =>
-            logger.info(s"Recovered $count upserts for index ${index.indexName}")
+        .batch(ServerService.RecoveryMaxBulkSize, json => Map(getEntityId(json) -> json)) {
+          (idMap, json) =>
+            val id = getEntityId(json)
+            val newJson = idMap.get(id) match {
+              case Some(oldJson) => {
+                logger.debug(
+                  s"Merging upserts ${getUpsertId(oldJson)} and ${getUpsertId(json)} for id $id"
+                )
+                oldJson.deepMerge(json)
+              }
+              case None => json
+            }
+            idMap.updated(id, newJson)
```

Comment:
New logic here to:
1. Enforce a max batch size
2. Merge upserts in-memory when possible

The default max-batch-size enforcer isn't smart enough to notice when two records merge, but there's a "cost function" parameter we could tweak so that it doesn't increment the current-total-batch-size when we merge entries in the map instead of adding a new entry.

Reply:
No luck with the cost function, it's only allowed to take the next element (vs. the next element and the existing batch aggregation). I don't think there's any harm in being conservative with the batch size.